### PR TITLE
Some demangler changes

### PR DESCRIPTION
**Pull request by eeckstein**
* Support the final mangling prefix _$s
* Also accept symbol names without the $ prefix

For details see the commit messages

rdar://problem/37681432
rdar://problem/38546308

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@jrose-apple Are you fine with requiring -stdin for swift-demangle to read from stdin?
(see the commit message for the motivation)

**Comment by compnerd** [comment]
@eeckstein not sure if my opinion matters, but, I think that requiring the flag is a bit surprising.  `c++filt` (and `llvm-c++filt`) both default to that without any flags and will run a loop to read in input and decode each line.  This is extremely helpful for use in scripts and just generally when debugging things without a debugger helping you to undecorate the names.

**Review by jrose-apple — DISMISSED** [review]
I'm with @compnerd. `swift-demangle` is supposed to act like `c++filt` first and foremost; passing arguments on the command line is the convenient extra feature, not the original purpose.

**Comment by eeckstein** [comment]
no problem. I'll remove that change

**Comment by eeckstein** [comment]
@jrose-apple I removed that commit

**Comment by eeckstein** [comment]
@swift-ci smoke test

File: tools/swift-demangle/swift-demangle.cpp [hunk 1]
```diff
@@ -175,7 +175,7 @@ static int demangleSTDIN(const swift::Demangle::DemangleOptions &options) {
   // This doesn't handle Unicode symbols, but maybe that's okay.
   // Also accept the future mangling prefix.
   // TODO: remove the "_S" as soon as MANGLING_PREFIX_STR gets "_S".
-  llvm::Regex maybeSymbol("(_T|_*\\$S|" MANGLING_PREFIX_STR ")[_a-zA-Z0-9$.]+");
+  llvm::Regex maybeSymbol("_(T|\\$[Ss]|" MANGLING_PREFIX_STR ")[_a-zA-Z0-9$.]+");
```
Comment (jrose-apple) [inline tools/swift-demangle/swift-demangle.cpp]: @graydon and I just noticed that `MANGLING_PREFIX_STR` will have an unescaped `$` in it, which means that things *without* the leading underscore weren't getting matched. Moving the underscore to the outside doesn't fix that either.

Comment (jrose-apple) [reply tools/swift-demangle/swift-demangle.cpp]: Maybe `"(_T|_?\\$[Ss])"` and leave out the `MANGLING_PREFIX_STR` part entirely?

Comment (jrose-apple) [reply tools/swift-demangle/swift-demangle.cpp]: Well, you didn't regress it. :-)

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**